### PR TITLE
repository, migrations, settings added

### PR DESCRIPTION
**Pull request by Kalasouski**

### CODE REVIEW

File: cmd/main.go [hunk 1]
```diff
@@ -1,47 +1,29 @@
...
 	"github.com/gin-gonic/gin"
 	_ "github.com/gin-gonic/gin"
-	"github.com/jmoiron/sqlx"
+	"github.com/joho/godotenv"
 	_ "github.com/lib/pq"
-	"os"
+	"log"
+	"userservice/internal/handlers"
+	"userservice/internal/repository"
+	"userservice/internal/settings"
```
Comment (DrGermanius) [inline cmd/main.go:11]: please use goimports for sorting(u can set up it in your ide)

File: internal/migrations/20220804091629_userdb.sql [hunk 2]
```diff
@@ -0,0 +1,16 @@
+-- +goose Up
+-- +goose StatementBegin
+CREATE TABLE users
+(
+    Id      serial primary key,
+    Name    varchar(32),
+    Phone   varchar(10) unique,
+    Email   varchar(32),
```
Comment (DrGermanius) [inline internal/migrations/20220804091629_userdb.sql:8]: email must be unique too

File: internal/migrations/20220804091629_userdb.sql [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    Password    varchar(256)
```
Comment (DrGermanius) [inline internal/migrations/20220804091629_userdb.sql:9]: please add CreatedAt, UpdatedAt fields

File: internal/migrations/migration.go [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+	"github.com/pressly/goose/v3"
+)
+
+func UpUserDb(db *sql.DB) error {
+	err := goose.Up(db, "internal/migrations")
+	if err != nil {
+		return err
+	}
+	return nil
+}
```
Comment (DrGermanius) [inline internal/migrations/migration.go:14]: is it will work inside of docker container?

File: internal/settings/settings.go [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+func GetDBSettings() (*DBSettings, error) {
+	setting := &DBSettings{
+		Host:     os.Getenv("HOST"),
+		Port:     os.Getenv("DBPORT"),
+		User:     os.Getenv("DBUSER"),
+		Password: os.Getenv("PASSWORD"),
+		DBName:   os.Getenv("DBNAME"),
+		SSLMode:  os.Getenv("SSLMODE")}
```
Comment (DrGermanius) [inline internal/settings/settings.go:23]: what will happen if one of these envs is not set?

**Review by DrGermanius — CHANGES_REQUESTED** [review]

File: internal/migrations/migration.go (hunk 4 above)
Comment (Kalasouski) [reply internal/migrations/migration.go:14]: yes, it works

File: cmd/main.go [hunk 6]
```diff
@@ -1,47 +1,40 @@
...
+		Repo:        repo,
+		AppSettings: appSettings,
+	}
 
-	server.Run(":8080")
-}
\ No newline at end of file
+	router := gin.Default()
+	handlers.InitHandlers(router, service)
+	err = router.Run(":8080")
```
Comment (DrGermanius) [inline cmd/main.go:36]: let's move this port to appSettings

File: internal/handlers/handlers.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+	user := models.LoginUser{}
+	err := ctx.BindJSON(&user)
+	if err != nil {
+		ctx.JSON(400, gin.H{
+			"message": "Invalid provided json"})
+		return
+	}
+	token, err := h.Service.Login(user)
+	if err != nil {
+		ctx.JSON(400, gin.H{
```
Comment (DrGermanius) [inline internal/handlers/handlers.go:75]: Why 400 ?

File: internal/migrations/20220808115101_userdb.sql [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+BEGIN
+    NEW.UpdatedAt = NOW();
+    RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE TRIGGER users_UpdatedAtTime
+    BEFORE UPDATE ON users
+    FOR EACH ROW
+    EXECUTE PROCEDURE trigger_set_timestamp();
```
Comment (DrGermanius) [inline internal/migrations/20220808115101_userdb.sql:25]: where did you find that ?

**Review by DrGermanius — CHANGES_REQUESTED** [review]

File: internal/migrations/20220808115101_userdb.sql (hunk 8 above)
Comment (Kalasouski) [reply internal/migrations/20220808115101_userdb.sql:25]: > where did you find that ?
I googled "postgesql updatedat triggers"

File: internal/handlers/handlers.go (hunk 7 above)
Comment (Kalasouski) [reply internal/handlers/handlers.go]: Fixed in new commit. I left 400 where I'm not sure which code is more appropriate

File: internal/handlers/handlers.go [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+	authorized.Use(h.tokenMiddleware)
+	authorized.GET("/getRating", h.ratingHandler)
+}
+
+func (h Handler) tokenMiddleware(ctx *gin.Context) {
+	token, err := extractToken(ctx)
+	if err != nil {
+		ctx.JSON(http.StatusUnauthorized, gin.H{
+			"message": fmt.Sprintf("Something went wrong during token extracting: %v", err)})
+		ctx.Abort()
```
Comment (DrGermanius) [inline internal/handlers/handlers.go:35]: investigate how to remove gin.H{} from responses.

File: internal/handlers/handlers.go [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+func InitHandlers(server *gin.Engine, s *services.Service) {
+	h := Handler{Server: server, Service: s}
+	server.POST("/register", h.registerHandler)
+	server.POST("/login", h.loginHandler)
+	authorized := server.Group("/api")
+	authorized.Use(h.tokenMiddleware)
+	authorized.GET("/getRating", h.ratingHandler)
+}
+
+func (h Handler) tokenMiddleware(ctx *gin.Context) {
```
Comment (DrGermanius) [inline internal/handlers/handlers.go:30]: let's move it to middlewares package or file

File: internal/handlers/handlers.go [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+	token, err := h.Service.Register(user)
+	if err != nil {
+		if errors.Unwrap(err) == usererrors.ErrUserExists {
+			ctx.JSON(http.StatusConflict, gin.H{
+				"message": "Cannot insert new record to the db. User already exists"})
+			return
+		}
+		ctx.JSON(http.StatusBadRequest, gin.H{
+			"message": "Cannot insert new record to the db."})
+		return
```
Comment (DrGermanius) [inline internal/handlers/handlers.go:67]: please create a function with mapping "your errors -> status codes"

File: internal/models/auth.go [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+	Phone string `json:"phone"`
+	Email string `json:"email"`
+}
+
+type RegisterUser struct {
+	Name     string `json:"name"`
+	Phone    string `json:"phone"`
+	Email    string `json:"email"`
+	Password string `json:"password"`
+}
```
Comment (DrGermanius) [inline internal/models/auth.go:16]: please add validation tags

File: internal/services/service.go [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+
+	if err != nil {
+		return nil, err
+	}
+	return &Token{token}, nil
+}
+
+func (s Service) ParseToken(token *Token) (jwt.Claims, error) {
+	parsedToken, err := jwt.ParseWithClaims(token.Token, &models.CustomClaims{}, func(token *jwt.Token) (interface{}, error) {
+		key := settings.GetenvDefault("JWT_SECRET", "my-32-character-ultra-secure-and-ultra-long-secret")
```
Comment (DrGermanius) [inline internal/services/service.go:91]: meh

**Review by DrGermanius — CHANGES_REQUESTED** [review]

File: internal/handlers/auth_handlers.go [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+import (
+	"fmt"
+	"github.com/gin-gonic/gin"
+	"net/http"
+	"strings"
+	"userservice/internal/models"
+	"userservice/internal/responses"
+	"userservice/internal/services"
+	"userservice/internal/usererrors"
+)
```
Comment (DrGermanius) [inline internal/handlers/auth_handlers.go:12]: please use "go imports"

File: internal/handlers/auth_handlers.go [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+	err := ctx.BindJSON(&user)
+	if err != nil {
+		ctx.JSON(http.StatusUnprocessableEntity,
+			responses.NewServerRequestError(fmt.Sprintf("Invalid provided json: %v", err)))
+		return
+	}
+	token, err := h.Service.Register(user)
+	if err != nil {
+		ctx.JSON(usererrors.ErrorToCode(err),
+			responses.NewServerRequestError(fmt.Sprintf("Something went wrong during registering: %v", err)))
```
Comment (DrGermanius) [inline internal/handlers/auth_handlers.go:25]: very verbose

File: internal/handlers/handlers.go [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+	h := Handler{Server: server, Service: s}
+	server.POST("/register", h.registerHandler)
+	server.POST("/login", h.loginHandler)
+	authorized := server.Group("/api")
+	authorized.Use(h.tokenMiddleware)
+	authorized.GET("/getRating", h.ratingHandler)
+}
+
+func (h Handler) ratingHandler(ctx *gin.Context) {
+	claims, _ := ctx.Get("Claims")
```
Comment (DrGermanius) [inline internal/handlers/handlers.go:30]: err ignoring?

File: internal/handlers/middlewares.go [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+	"userservice/internal/usererrors"
+)
+
+func (h Handler) tokenMiddleware(ctx *gin.Context) {
+	token, err := extractToken(ctx)
+	if err != nil {
+		ctx.JSON(usererrors.ErrorToCode(err),
+			responses.NewServerUnauthorizedResponse(fmt.Sprintf("Something went wrong during token extracting: %v", err)),
+		)
+		ctx.Abort()
```
Comment (DrGermanius) [inline internal/handlers/middlewares.go:16]: i believe you don't need to do ctx.Abort()

File: internal/usererrors/errors.go [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+	errorCodes[ErrWrongClaims] = http.StatusUnprocessableEntity
+	errorCodes[ErrInvalidToken] = http.StatusUnauthorized
+	errorCodes[ErrWrongPass] = http.StatusUnauthorized
+	errorCodes[ErrTokenExpired] = http.StatusUnauthorized
+	errorCodes[ErrInternal] = http.StatusInternalServerError
+}
+
+func ErrorToCode(e error) int {
+	if code, ok := errorCodes[e]; !ok {
+		return http.StatusBadRequest
```
Comment (DrGermanius) [inline internal/usererrors/errors.go:36]: better to return 500

File: internal/usererrors/errors.go [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+	errorCodes[ErrUserExists] = http.StatusConflict
+	errorCodes[ErrUserNotFound] = http.StatusNotFound
+	errorCodes[ErrNoAuthHeader] = http.StatusUnauthorized
+	errorCodes[ErrNoBearerPrefix] = http.StatusUnauthorized
+	errorCodes[ErrWrongClaims] = http.StatusUnprocessableEntity
+	errorCodes[ErrInvalidToken] = http.StatusUnauthorized
+	errorCodes[ErrWrongPass] = http.StatusUnauthorized
+	errorCodes[ErrTokenExpired] = http.StatusUnauthorized
+	errorCodes[ErrInternal] = http.StatusInternalServerError
+}
```
Comment (DrGermanius) [inline internal/usererrors/errors.go:32]: you can just init a map

var m := map[aaa]vvv {
.. values ....
 }

File: internal/services/service.go [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+		return []byte(s.AppSettings.JwtSecret), nil
+	})
+
+	if err != nil {
+		v, _ := err.(*jwt.ValidationError)
+		if v.Errors == jwt.ValidationErrorExpired {
+			return nil, usererrors.ErrTokenExpired
+		}
+		return nil, usererrors.ErrInvalidToken
+	}
```
Comment (DrGermanius) [inline internal/services/service.go:99]: finally you lose your err and you don't know what happens

File: internal/handlers/middlewares.go (hunk 17 above)
Comment (Kalasouski) [reply internal/handlers/middlewares.go:18]: If I remove abort call then thenext handler function will be executed

Comment (DrGermanius) [reply internal/handlers/middlewares.go:18]: you're right
let's use the c.AbortWithStatus() or AbortWIthJSON instead. Looks more clear

**Review by DrGermanius — CHANGES_REQUESTED** [review]

**Pull request merged**